crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,557 @@
+//! A buffered implementation that automatically groups native prices API
+//! requests into batches.
+
+use {
+    crate::price_estimation::{
+        native::{NativePriceEstimateResult, NativePriceEstimating},
+        PriceEstimationError,
+    },
+    anyhow::anyhow,
+    async_trait::async_trait,
+    futures::{
+        channel::mpsc,
+        future::FutureExt as _,
+        stream::{self, FusedStream, Stream, StreamExt as _},
+    },
+    primitive_types::H160,
+    std::{
+        collections::{HashMap, HashSet},
+        future::Future,
+        num::NonZeroUsize,
+        sync::Arc,
+        time::Duration,
+    },
+    tokio::{sync::broadcast, task::JoinHandle, time::error::Elapsed},
+};
+
+/// Buffered configuration.
+#[derive(Clone)]
+#[allow(dead_code)]
+pub struct Configuration {
+    /// The maximum amount of concurrent batches to request.
+    ///
+    /// Specifying `None` means no limit on concurrency.
+    pub max_concurrent_requests: Option<NonZeroUsize>,
+    /// The maximum batch size.
+    pub max_batch_len: usize,
+    /// An additional minimum delay to wait for collecting requests.
+    ///
+    /// The delay starts counting after receiving the first request.
+    pub batch_delay: Duration,
+    /// The timeout to wait for the result to be ready
+    pub result_ready_timeout: Duration,
+    /// Maximum capacity of the broadcast channel to store the native prices
+    /// results
+    pub broadcast_channel_capacity: usize,
+}
+
+/// Trait for fetching a batch of native price estimates.
+#[allow(dead_code)]
+#[mockall::automock]
+#[async_trait]
+pub trait NativePriceBatchFetching: Sync + Send + NativePriceEstimating {
+    /// Fetches a batch of native price estimates.
+    ///
+    /// It returns a HashMap which maps the token with its native price
+    /// estimator result
+    async fn fetch_native_prices(
+        &self,
+        tokens: &HashSet<H160>,
+    ) -> Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError>;
+}
+
+#[async_trait]
+impl NativePriceEstimating for MockNativePriceBatchFetching {
```

Comment:
Yeah, ideally. But `mockall` does create code not only for test, but in general :/

Reply:
Huh, is this only necessary now that the new trait extends the original native price trait and mockall can't handle that well?
Seems interesting and a bit unfortunate. 😅 
Depending on how widely you actually need the mock struct you could also just apply the mock macro for testing purpose with: `#[cfg_attr(test, mockall::automock)]`. That would allow you to move this implementation into the test section.

Reply:
sweet, thank you! we should propagate it to all the mockalls when possible, to avoid extra code in production :thinking:

Reply:
I tried that in the past but the issue is that you can't use this trick if you have to export the `Mock` struct outside of the crate which for example happens for other price estimation tests.
Also this code doesn't make a huge difference since it should get stripped out of the binary anyway since it's not used outside of tests.

Reply:
makes sense!